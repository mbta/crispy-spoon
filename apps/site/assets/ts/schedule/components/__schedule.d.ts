import { TypedRoutes } from "../../stop/components/__stop";
import { Route } from "../../__v3api";

export interface SchedulePageData {
  connections: TypedRoutes[];
  pdfs: SchedulePDF[];
  teasers: string | null;
  hours: string;
  fares: Fare[];
  fare_link: string;
  holidays: Holiday[];
  route: Route;
  schedule_note: ScheduleNote | null;
  stops: SimpleStop[];
}

export interface ScheduleNote {
  peak_service: string;
  offpeak_service: string;
  exceptions: ServiceException[];
}

export interface ServiceException {
  type: string;
  service: string;
}

export interface SimpleStop {
  id: string;
  name: string;
}

export interface SchedulePDF {
  title: string;
  url: string;
}

export interface Fare {
  title: string;
  price: string;
}

export interface Holiday {
  name: string;
  date: string;
}
