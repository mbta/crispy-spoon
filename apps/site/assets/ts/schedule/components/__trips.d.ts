import { Route, Trip, Stop, PredictedOrScheduledTime } from "../../__v3api";

export interface Journey {
  trip: Trip;
  route: Route;
  departure: Departure;
}

export interface EnhancedJourney extends Journey {
  realtime: PredictedOrScheduledTime;
}

export interface Departure {
  time: string;
  schedule: Schedule | null;
  prediction: Prediction | null;
}

export interface TripDeparture {
  schedule: TripSchedule;
  prediction: Prediction | null;
}

export interface TripInfo {
  times: TripDeparture[];
  vehicle: Vehicle | null;
  vehicle_stop_name: string;
  stop_count: number;
  status: string;
  duration: number;
  fare: Fare;
}

export interface Prediction {
  delay: number;
  status: string | null;
  track: string | null;
}

export interface Schedule {
  time: string;
  stop_sequence: number;
  pickup_type: number;
  "flag?": boolean;
  "early_departure?": boolean;
  "last_stop?": boolean;
}

export interface TripSchedule extends Schedule {
  stop: Stop;
  fare: Fare;
}

export interface Vehicle {
  trip_id: string;
  stop_id: string;
  status: string;
}

export interface Fare {
  price: string;
  fare_link: string;
}
