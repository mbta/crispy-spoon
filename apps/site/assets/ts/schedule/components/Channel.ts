import { Socket, Channel } from "phoenix";

declare global {
  interface Window {
    channels: { [id: string]: Channel };
    socket?: Socket;
  }
}

type UpdateEventName = "reset" | "add" | "update" | "setChannel";

interface UpdateEvent<T> {
  event: UpdateEventName;
  data: T;
}

interface RemoveEvent {
  event: "remove";
  data: string[];
}

export type SocketEvent<T> = UpdateEvent<T> | RemoveEvent;

export const doInitChannel = (id: string): Channel => {
  if (!window.socket) {
    window.socket = new Socket("/socket", {});
    window.socket.connect();
  }

  if (!window.channels) {
    window.channels = {};
  }

  if (window.channels[id]) return window.channels[id];

  const channel = window.socket.channel(id, {});
  window.channels[id] = channel;

  channel
    .join()
    .receive("error", ({ reason }) =>
      // eslint-disable-next-line no-console
      console.error(`failed to join ${id}`, reason)
    )
    // eslint-disable-next-line no-console
    .receive("timeout", () => console.error(`timeout joining ${id}`))
    .receive("ok", () => {
      if (id !== "vehicles:remove") {
        const route = id.split(":");
        channel.push("init", {
          // eslint-disable-next-line camelcase
          route_id: route[1],
          // eslint-disable-next-line camelcase
          direction_id: route[2]
        });
      }
    });

  return channel;
};

export const initChannel = <T>(
  channelId: string,
  onData: (data: SocketEvent<T>) => void
): void => {
  const channel = doInitChannel(channelId);
  channel.on("data", onData);
};

// NB: this removes *all* handlers from the channel. We're currently throwing
// away the reference returned by `channel.on`, so we can't remove a specific
// handler, but we don't have a use case for that yet.
export const stopChannel = (id: string): void => {
  if (window.channels && window.channels[id]) {
    window.channels[id].off("data");
  }
};
