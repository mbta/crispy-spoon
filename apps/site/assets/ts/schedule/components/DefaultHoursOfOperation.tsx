import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";

const DefaultHoursOfOperation = ({
  hours
}: {
  hours: string;
}): ReactElement<HTMLElement> | null =>
  hours ? (
    <ExpandableBlock
      header={{ text: "Hours of Operation", iconSvgText: null }}
      initiallyExpanded={false}
      id="hours"
    >
      <div
        className="m-schedule-page__sidebar-hours"
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{ __html: hours }}
      />
    </ExpandableBlock>
  ) : null;

export default DefaultHoursOfOperation;
