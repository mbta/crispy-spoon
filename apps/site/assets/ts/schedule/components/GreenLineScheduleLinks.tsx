import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";
import renderFa from "../../helpers/render-fa";
import { SchedulePDF } from "./__schedule";

const pdfLink = (pdf: SchedulePDF): ReactElement<HTMLElement> => (
  <a key={pdf.url} href={pdf.url} rel="noopener noreferrer" target="_blank">
    Open subway schedule PDF {renderFa("fa-arrow-up-right-from-square", "")}
  </a>
);

const getGreenLineScheduleLinks = (): ReactElement<HTMLElement> => {
  return (
    <>
      <div>
        <a href="/schedules/Green-B/line">B Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-C/line">C Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-D/line">D Line Schedule</a>
      </div>
      <div>
        <a href="/schedules/Green-E/line">E Line Schedule</a>
      </div>
    </>
  );
};

const GreenLineScheduleLinks = ({
  pdfs
}: {
  pdfs: SchedulePDF[];
}): ReactElement<HTMLElement> | null => {
  console.log(pdfs);
  return (
    <>
      <ExpandableBlock
        header={{ text: "Weekday Schedule", iconSvgText: null }}
        initiallyExpanded={false}
        id="weekday-hours"
      >
        <div className="m-schedule-page__sidebar-hours">
          <div className="fs-14 pt-14">Trains every 6-12 minutes</div>
          {getGreenLineScheduleLinks()}
        </div>
      </ExpandableBlock>
      <ExpandableBlock
        header={{ text: "Weekend Schedule", iconSvgText: null }}
        initiallyExpanded={false}
        id="weekend-hours"
      >
        <div className="m-schedule-page__sidebar-hours">
          <div className="font-weight-bold fs-18 pb-14">Saturday</div>
          <div className="fs-14 pt-14">Trains every 6-12 minutes</div>
          {getGreenLineScheduleLinks()}
          <div
            style={{
              borderBottomWidth: "0.5px",
              borderBottomColor: "black",
              borderBottomStyle: "solid"
            }}
            className="pt-18"
          ></div>
          <div className="font-weight-bold fs-18 pt-18 pb-14">Sunday</div>
          {getGreenLineScheduleLinks()}
          {<div className="fs-14 pt-18">{pdfLink(pdfs[0])}</div>}
        </div>
      </ExpandableBlock>
    </>
  );
};

export default GreenLineScheduleLinks;
