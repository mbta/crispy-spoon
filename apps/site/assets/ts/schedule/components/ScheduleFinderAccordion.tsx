import React, { ReactElement } from "react";
import icon from "../../../static/images/icon-schedule-finder.svg";
import ExpandableBlock from "../../components/ExpandableBlock";
import ScheduleFinder from "./ScheduleFinder";
import {
  DirectionId,
  EnhancedRoute,
  ServiceWithServiceDate
} from "../../__v3api";
import { SimpleStopMap, RoutePatternsByDirection } from "./__schedule";

interface Props {
  directionId: DirectionId;
  route: EnhancedRoute;
  stops: SimpleStopMap;
  services: ServiceWithServiceDate[];
  ratingEndDate: string;
  routePatternsByDirection: RoutePatternsByDirection;
}

const ScheduleFinderAccordion = ({
  directionId,
  route,
  stops,
  services,
  ratingEndDate,
  routePatternsByDirection
}: Props): ReactElement<HTMLDivElement> => (
  <div className="schedule-finder--mobile">
    <ExpandableBlock
      header={{ text: "Schedule Finder", iconSvgText: icon }}
      initiallyExpanded={false}
      id="schedule-finder-mobile"
      panelClassName="schedule-finder--accordion"
    >
      <ScheduleFinder
        directionId={directionId}
        hideHeader
        route={route}
        stops={stops}
        services={services}
        ratingEndDate={ratingEndDate}
        routePatternsByDirection={routePatternsByDirection}
      />
    </ExpandableBlock>
  </div>
);

export default ScheduleFinderAccordion;
