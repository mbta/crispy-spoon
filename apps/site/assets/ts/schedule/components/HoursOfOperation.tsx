import React, { ReactElement } from "react";
import { isGreenLine, isRapidTransit } from "../../models/route";
import { EnhancedRoute } from "../../__v3api";
import DefaultHoursOfOperation from "./DefaultHoursOfOperation";
import GreenLineScheduleLinks from "./GreenLineScheduleLinks";
import RapidTransitHoursOfOperation from "./RapidTransitHoursOfOperation";
import { SchedulePDF, ScheduleNote } from "./__schedule";

const HoursOfOperation = ({
  route,
  pdfs,
  hours,
  scheduleNote
}: {
  route: EnhancedRoute;
  pdfs: SchedulePDF[];
  hours: string;
  scheduleNote: ScheduleNote | null;
}): ReactElement<HTMLElement> | null => {
  if (isGreenLine(route)) {
    return <GreenLineScheduleLinks pdfs={pdfs} />;
  }
  if (isRapidTransit(route)) {
    return (
      <RapidTransitHoursOfOperation
        pdfs={pdfs}
        route={route}
        scheduleNote={scheduleNote}
      />
    );
  }
  return <DefaultHoursOfOperation hours={hours} />;
};

export default HoursOfOperation;
