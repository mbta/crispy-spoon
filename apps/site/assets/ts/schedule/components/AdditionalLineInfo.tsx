import React, { ReactElement } from "react";
import { Holiday, SchedulePDF, Fare, ScheduleNote } from "./__schedule";
import { EnhancedRoute } from "../../__v3api";
import { TypedRoutes } from "../../stop/components/__stop";
import PDFSchedules from "./PDFSchedules";
import Connections from "./Connections";
import ContentTeasers from "./ContentTeasers";
import Fares from "./Fares";
import UpcomingHolidays from "./UpcomingHolidays";
import HoursOfOperation from "./HoursOfOperation";
import { isSubwayRoute } from "../../models/route";

interface Props {
  teasers: string | null;
  pdfs: SchedulePDF[];
  connections: TypedRoutes[];
  fares: Fare[];
  fareLink: string;
  route: EnhancedRoute;
  holidays: Holiday[];
  scheduleNote: ScheduleNote | null;
  hours: string;
}

const AdditionalLineInfo = ({
  connections,
  pdfs,
  teasers,
  fares,
  holidays,
  fareLink,
  route,
  hours,
  scheduleNote
}: Props): ReactElement<HTMLElement> => (
  <>
    <ContentTeasers teasers={teasers} />
    <PDFSchedules pdfs={pdfs} />
    <Connections connections={connections} />
    <Fares fares={fares} fareLink={fareLink} routeType={route.type} />
    {/* Only show the hours for non subway lines */}
    {!isSubwayRoute(route) && (
      <HoursOfOperation
        route={route}
        pdfs={pdfs}
        scheduleNote={scheduleNote}
        hours={hours}
      />
    )}
    <UpcomingHolidays holidays={holidays} />
  </>
);

export default AdditionalLineInfo;
