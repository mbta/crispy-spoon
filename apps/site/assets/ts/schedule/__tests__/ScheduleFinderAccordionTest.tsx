import React from "react";
import renderer from "react-test-renderer";
import { createReactRoot } from "../../app/helpers/testUtils";
import ScheduleFinderAccordion from "../components/ScheduleFinderAccordion";
import { EnhancedRoute, ServiceWithServiceDate } from "../../__v3api";
import { ServiceSchedule } from "../components/__schedule";

const services: ServiceWithServiceDate[] = [
  {
    valid_days: [1, 2, 3, 4, 5],
    typicality: "typical_service",
    type: "weekday",
    start_date: "2019-07-02",
    service_date: "2019-07-09",
    removed_dates_notes: { "2019-07-04": "Independence Day" },
    removed_dates: ["2019-07-04"],
    name: "Weekday",
    id: "BUS319-O-Wdy-02",
    end_date: "2019-08-30",
    description: "Weekday schedule",
    added_dates_notes: {},
    added_dates: []
  },
  {
    valid_days: [6],
    typicality: "typical_service",
    type: "saturday",
    start_date: "2019-07-06",
    service_date: "2019-07-09",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Saturday",
    id: "BUS319-P-Sa-02",
    end_date: "2019-08-31",
    description: "Saturday schedule",
    added_dates_notes: {},
    added_dates: []
  },
  {
    valid_days: [7],
    typicality: "typical_service",
    type: "sunday",
    start_date: "2019-07-07",
    service_date: "2019-07-09",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Sunday",
    id: "BUS319-Q-Su-02",
    end_date: "2019-08-25",
    description: "Sunday schedule",
    added_dates_notes: {},
    added_dates: []
  }
] as ServiceWithServiceDate[];

/* eslint-disable @typescript-eslint/camelcase */
const body = '<div id="react-root"></div>';
const route: EnhancedRoute = {
  alert_count: 0,
  description: "",
  direction_destinations: { 0: "Oak Grove", 1: "Forest Hills" },
  direction_names: { 0: "Inbound", 1: "Outbound" },
  header: "",
  id: "Orange",
  long_name: "Orange Line",
  name: "Orange",
  type: 1
};
const stops = [
  {
    name: "SL",
    id: "741",
    is_closed: false,
    zone: "1"
  },
  {
    name: "Abc",
    id: "123",
    is_closed: false,
    zone: null
  },
  {
    name: "Wellington",
    id: "place-welln",
    is_closed: true,
    zone: null
  }
];

it("renders", () => {
  createReactRoot();
  const tree = renderer
    .create(
      <ScheduleFinderAccordion
        route={route}
        stops={stops}
        directionId={0}
        services={services}
        serviceSchedules={{}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
