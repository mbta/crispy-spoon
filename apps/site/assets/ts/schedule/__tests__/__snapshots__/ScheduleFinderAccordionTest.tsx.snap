// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="schedule-finder--mobile"
>
  <h3
    aria-controls="panel-schedule-finder-mobile"
    aria-expanded={false}
    className="c-expandable-block__header"
    id="header-schedule-finder-mobile"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      aria-hidden="true"
      className="notranslate c-expandable-block__header-icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
    <div
      className="c-expandable-block__header-text"
    >
      Schedule Finder
      <span
        className="c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF107;",
          }
        }
      />
    </div>
  </h3>
  <noscript>
    <style>
      #header-schedule-finder-mobile { display: none; }
    </style>
    <h3
      className="c-expandable-block__header"
    >
      <span
        aria-hidden="true"
        className="notranslate c-expandable-block__header-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Schedule Finder
      <span
        className="c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </h3>
    <div
      className="c-expandable-block__panel"
      role="region"
    >
      <div
        className="schedule-finder"
      >
        <div>
          Choose a stop to get schedule information and real-time departure predictions.
        </div>
        <div
          className="schedule-finder__select-container "
          id="sf_direction_select_container"
          onClick={[Function]}
          onKeyUp={[Function]}
          role="button"
          tabIndex={0}
        >
          <select
            className="schedule-finder__select"
            id="sf_direction_select"
            onChange={[Function]}
            onKeyUp={[Function]}
            value=""
          >
            <option
              value=""
            >
              Choose a direction
            </option>
            <option
              value="0"
            >
              INBOUND
               
              Oak Grove
            </option>
            <option
              value="1"
            >
              OUTBOUND
               
              Forest Hills
            </option>
          </select>
          <span
            aria-hidden="true"
            className="notranslate c-svg__icon schedule-finder__arrow"
            dangerouslySetInnerHTML={
              Object {
                "__html": "SVG",
              }
            }
          />
        </div>
        <div
          className="schedule-finder__select-container "
          id="sf_origin_select_container"
          onClick={[Function]}
          onKeyUp={[Function]}
          role="button"
          tabIndex={0}
        >
          <select
            className="schedule-finder__select schedule-finder__select--noclick"
            id="sf_origin_select"
            onChange={[Function]}
            onKeyUp={[Function]}
            value=""
          >
            <option
              value=""
            >
              Choose an origin stop
            </option>
            <option
              value="place-welln"
            >
              Wellington
            </option>
            <option
              value="123"
            >
              Abc
            </option>
            <option
              value="741"
            >
              SL
            </option>
          </select>
          <span
            aria-hidden="true"
            className="notranslate c-svg__icon schedule-finder__arrow"
            dangerouslySetInnerHTML={
              Object {
                "__html": "SVG",
              }
            }
          />
        </div>
        <div
          className="text-right"
        >
          <input
            className="btn btn-primary"
            onClick={[Function]}
            type="submit"
            value="Get schedules"
          />
        </div>
      </div>
    </div>
  </noscript>
</div>
`;
